ocaml/typing/ctype.ml

```diff
@@ -2195,137 +2125,107 @@ let type_jkind_sub env ty jkind =
 
-(* The ~fixed argument controls what effects this may have on `ty`.  If false,
-   then we will update the jkind of type variables to make the check true, if
-   possible.  If true, we won't (but will still instantiate sort variables).
+    (* Handle the [Tpoly] case out here so [Tvar]s wrapped in [Tpoly]s can get
+       the treatment above. *)
+    | Tpoly (t, _) -> loop ~fuel ~expanded:false t ty's_jkind jkind
 
-   Precondition: [jkind] is not [any]. This common case is short-circuited
-   before calling this function. (Though the current implementation is still
-   correct on [any].)
-*)
-let constrain_type_jkind ~fixed env ty jkind =
-  let ty_jkind, result = type_jkind_sub env ty jkind in
-  match result with
-  | Success -> Ok ()
-  | Type_vars tvs ->
-    let constrain_one_var { estimate; bound; ty } =
-      if fixed then Jkind.sub_or_error estimate bound else
-      let jkind_inter =
-        Jkind.intersection_or_error ~reason:Tyvar_refinement_intersection
-          estimate bound
-      in
-      Result.map (set_var_jkind ty) jkind_inter
-    in
-    Misc.Stdlib.List.iter_until_error ~f:constrain_one_var tvs
-  | Missing_cmi missing_cmi ->
-    Error Jkind.(Violation.of_ ~missing_cmi
-      (Not_a_subjkind
-         (History.update_reason ty_jkind (Missing_cmi missing_cmi), jkind)))
-  | Failure _ ->
-    Error (Jkind.Violation.of_ (Not_a_subjkind (ty_jkind, jkind)))
+    | _ ->
+       match Jkind.sub_or_intersect ty's_jkind jkind with
+       | Sub -> Ok ()
+       | Disjoint ->
+          (* Reporting that [ty's_jkind] must be a subjkind of [jkind] is not
+             always right.  Suppose we had [type ('a : wprd) t = 'a] and we were
+             checking ['a t] against [value]. Then it would be enough for [word]
+             and [value] to have an intersection, not for one to be a subjkind
+             of another. But getting this reporting correct would require
+             arbitrary amounts of expansion and looking through [@@unboxed]
+             types. So we don't, settling for the slightly worse error
+             message. *)
+          Error (Jkind.Violation.of_ (Not_a_subjkind (ty's_jkind, jkind)))
+       | Has_intersection ->
+          match get_desc ty with
+          | Tconstr _ ->
+             if not expanded
+             then
+               let ty = expand_head_opt env ty in
+               loop ~fuel ~expanded:true ty (estimate_type_jkind env ty) jkind
+             else
+               begin match unbox_once env ty with
+               | Missing path -> Error (Jkind.Violation.of_ ~missing_cmi:path
+                                          (Not_a_subjkind (ty's_jkind, jkind)))
+               | Final_result ->
+                  Error (Jkind.Violation.of_ (Not_a_subjkind (ty's_jkind, jkind)))
+               | Stepped ty ->
+                  loop ~fuel:(fuel - 1) ~expanded:false ty
+                    (estimate_type_jkind env ty) jkind
+               end
+          | Tunboxed_tuple ltys ->
```

Comment:
Great catch.

I decided to do this by a special case in `constrain_type_jkind` itself, not `decompose_product`, because decomposing an `any` on the left (in `ty's_jkind`) would be wrong. If we had left- and right-kinds, maybe I'd have done this differently, but it was easy enough to special-case in `constrain_type_jkind`, and I think the solution is sufficient.